src/icp_canister_backend/src/lib.rs

```diff
@@ -433,6 +463,80 @@ pub async fn slash(receiver: Principal, amount: Nat) -> Result<(), types::PoolEr
     Ok(())
 }
 
+#[ic_cdk::update]
+pub async fn reward_pool(amount: Nat) -> Result<(), types::PoolError> {
+    let ledger_principal = TOKEN_ID
+        .with(|cell| {
+            let stored = cell.borrow().get().clone();
+            Some(stored)
+        })
+        .ok_or(types::PoolError::LedgerNotSet)?;
+    
+    let reward_subaccount = get_reward_subaccount();
+    let from_account = types::Account {
+        owner: ic_cdk::api::id(),
+        subaccount: Some(reward_subaccount.to_vec()),
+    };
+    
+    let balance_result: Result<(Nat,), _> =
+        call(ledger_principal, "icrc1_balance_of", (from_account,)).await;
+    
+    let balance = match balance_result {
+        Ok((balance,)) => balance,
+        Err(_) => return Err(types::PoolError::LedgerCallFailed),
+    };
+    
+    let required_amount = amount.clone() + TRANSFER_FEE.clone();
+    if balance < required_amount {
+        return Err(types::PoolError::InsufficientBalance);
+    }
+   
+    let transfer_args = (types::TransferArg {
+        from_subaccount: Some(reward_subaccount.to_vec()),
+        to: types::Account {
+            owner: ic_cdk::api::id(),
+            subaccount: None,
+        },
+        amount: amount.clone(),
+        fee: Some(TRANSFER_FEE.clone()),
+        memo: None,
+        created_at_time: None,
+    },);
+    
+    let transfer_result: Result<(Result<Nat, types::TransferError>,), _> =
+        call(ledger_principal, "icrc1_transfer", transfer_args).await;
+    
+    if transfer_result.is_err() || transfer_result.as_ref().unwrap().0.is_err() {
+        return Err(types::PoolError::TransferFailed);
+    }
+    
+    let current_time = ic_cdk::api::time() / 1_000_000_000;
+    let reward_duration = 365 * 24 * 60 * 60; 
+    let last_reward_episode = (current_time + reward_duration) / EPISODE_DURATION;
+    
+    let reward_rate_increase = amount / Nat::from(reward_duration);
```

Comment:
reward rate increase should be calculated a little bit differently, because actual reward is tighted to the episode duration so 365 * 24 * 60 * 60 is minimum duration and to this amount should be added time to the current episode finished.